### PR TITLE
RD-2045 - Conversion of Labels feature files to TypeScript

### PR DESCRIPTION
**Pull request by qooban**
This PR is a preparation for RD-2045 and RD-2041. Just JS->TS conversion, no new functionality added, hopefully none of the already existing functionality broken 🙂 

System tests: https://jenkins.cloudify.co/view/UI%20Health%20View/job/Stage-UI-System-Test/576/

### CODE REVIEW

File: widgets/common/src/hooks/useLabelInput.ts [hunk 1]
```diff
@@ -1,23 +1,30 @@
+import { SyntheticEvent } from 'react';
+import type { DropdownProps } from 'semantic-ui-react';
+
 const allowedCharacters = /^[a-z][a-z0-9._-]*$/i;
 const maxInputLength = 256;
 
-function useLabelInput(onChange, { allowAnyValue = false, initialValue = '' }) {
+// TODO: Add proper typing for onChange function
+function useLabelInput(onChange: (value: any) => void, { allowAnyValue = false, initialValue = '' }) {
```
Comment (qooban) [inline widgets/common/src/hooks/useLabelInput.ts:8]: This hook needs to be refactored. I'll work on that when I'll be implementing RD-2045 and RD-2041.

**Comment by qooban** [comment]
@kubama To not overload you with the reviews, I requested review only from @Gelio right now. This is 100% JS-TS conversion, so I think we can do an exception and rely on one reviewer. Let me know if you see it differently. And of course, if you have time, feel free to join.

File: widgets/common/src/labels/ManageLabelsModal.tsx [hunk 2]
```diff
@@ -1,6 +1,9 @@
-import LabelsModal from './LabelsModal';
+import { FunctionComponent } from 'react';
+import LabelsModal, { LabelsModalProps } from './LabelsModal';
 
-export default function ManageLabelsModal(props) {
+type ManageLabelsModalProps = Omit<LabelsModalProps, 'i18nHeaderKey' | 'i18nApplyKey'>;
```
Comment (Gelio) [inline widgets/common/src/labels/ManageLabelsModal.tsx:4]: Nice use of `Omit` :+1:

**Review by Gelio — APPROVED** [review]
Looks good overall :+1: I have added some mostly minor comments

File: widgets/common/src/hooks/useLabelInput.ts [hunk 3]
```diff
@@ -1,23 +1,30 @@
...
+        submitChange: (_event: SyntheticEvent | null, data: DropdownProps) => {
             if (allowAnyValue) {
                 onChange(data.value);
                 resetInputValue();
                 return;
             }
 
             // supports both dropdown as well as regular input
-            const lowercasedNewTypedValue = _.toLower(data.searchQuery ?? data.value).substr(0, maxInputLength);
+            const lowercasedNewTypedValue = _.toLower((data.searchQuery as string) ?? (data.value as string)).substr(
```
[the remark this answers is not in the document]
Comment (qooban) [reply widgets/common/src/hooks/useLabelInput.ts]: Thanks for the explanation. I changed it according to yours suggestion.

File: widgets/common/src/hooks/useLabelInput.ts [hunk 4]
```diff
@@ -35,4 +42,15 @@ function useLabelInput(onChange, { allowAnyValue = false, initialValue = '' }) {
...
+declare global {
+    namespace Stage {
+        interface Hooks {
+            useLabelInput: typeof useLabelInput;
+        }
+    }
+}
+
+// NOTE: prevents leaking variables as global in TS
+export {};
```
[the remark this answers is not in the document]
Comment (qooban) [reply widgets/common/src/hooks/useLabelInput.ts]: Oh, good to know, so I removed it then.

File: widgets/common/src/labels/CommonDropdown.tsx [hunk 5]
```diff
@@ -1,8 +1,39 @@
...
+const CommonDropdown: FunctionComponent<CommonDropdownProps> = ({
+    allowAdditions = false,
+    innerRef = null,
+    onChange,
+    value = null,
+    ...rest
+}) => {
     const { useEffect } = React;
     const {
+        // @ts-ignore Property 'DynamicDropdown' does not exist on type 'typeof Common'
```
[the remark this answers is not in the document]
Comment (qooban) [reply widgets/common/src/labels/CommonDropdown.tsx]: Yep, that was automatically suggested by WebStorm, so I used it. Changed according to yours suggestion.

File: widgets/common/src/labels/CommonDropdown.tsx [hunk 6]
```diff
@@ -1,8 +1,39 @@
...
+
+interface CommonDropdownProps extends KeyAndValueDropdownProps {
+    additionLabel: ComponentProps<typeof Dropdown>['additionLabel'];
+    disabled?: ComponentProps<typeof Dropdown>['disabled'];
+    fetchUrl: string;
+    name: ComponentProps<typeof Dropdown>['name'];
+    multiple?: ComponentProps<typeof Dropdown>['multiple'];
+    noResultsMessage?: ComponentProps<typeof Dropdown>['noResultsMessage'];
+    placeholder: ComponentProps<typeof Dropdown>['placeholder'];
+    tabIndex: ComponentProps<typeof Dropdown>['tabIndex'];
```
[the remark this answers is not in the document]
Comment (qooban) [reply widgets/common/src/labels/CommonDropdown.tsx]: Good idea 👍🏻  Changed.

File: widgets/common/src/labels/CommonDropdown.tsx [hunk 7]
```diff
@@ -1,8 +1,39 @@
+import { ComponentProps, FunctionComponent, MutableRefObject } from 'react';
+import { Dropdown } from 'semantic-ui-react';
 import ValidationErrorPopup from './ValidationErrorPopup';
 
-export default function CommonDropdown({ innerRef, fetchUrl, onChange, toolbox, allowAdditions, value, ...rest }) {
+type CommonDropdownValue = string | string[];
+type CommonDropdownOnChange = ((value: string) => void) | ((value: string[]) => void);
```
[the remark this answers is not in the document]
Comment (qooban) [reply widgets/common/src/labels/CommonDropdown.tsx:6]: Unfortunately not. I tried it that way before, but there are problems then 

<details>
  <summary>Error log from `npm run check-types`</summary>

```
widgets/common/src/filters/inputs/LabelValueInput.tsx:45:25 - error TS2322: Type '(value: string[]) => void' is not assignable to type 'CommonDropdownOnChange'.
  Types of parameters 'value' and 'value' are incompatible.
    Type 'CommonDropdownValue' is not assignable to type 'string[]'.
      Type 'string' is not assignable to type 'string[]'.

45                         onChange={onValueChange}
                           ~~~~~~~~

  widgets/common/src/labels/CommonDropdown.tsx:11:5
    11     onChange: CommonDropdownOnChange;
           ~~~~~~~~
    The expected type comes from property 'onChange' which is declared here on type 'IntrinsicAttributes & ValueDropdownProps & { children?: ReactNode; }'

widgets/common/src/labels/KeyDropdown.tsx:4:11 - error TS2430: Interface 'KeyDropdownProps' incorrectly extends interface 'KeyAndValueDropdownProps'.
  Types of property 'onChange' are incompatible.
    Type '(value: string) => void' is not assignable to type 'CommonDropdownOnChange'.
      Types of parameters 'value' and 'value' are incompatible.
        Type 'CommonDropdownValue' is not assignable to type 'string'.
          Type 'string[]' is not assignable to type 'string'.

4 interface KeyDropdownProps extends KeyAndValueDropdownProps {
            ~~~~~~~~~~~~~~~~

widgets/common/src/labels/KeyDropdown.tsx:26:13 - error TS2322: Type '(value: string) => void' is not assignable to type 'CommonDropdownOnChange'.

26             onChange={onChange}
               ~~~~~~~~

  widgets/common/src/labels/CommonDropdown.tsx:11:5
    11     onChange: CommonDropdownOnChange;
           ~~~~~~~~
    The expected type comes from property 'onChange' which is declared here on type 'IntrinsicAttributes & CommonDropdownProps & { children?: ReactNode; }'

widgets/common/src/labels/LabelsInput.tsx:179:33 - error TS2322: Type 'Dispatch<SetStateAction<string>>' is not assignable to type 'CommonDropdownOnChange'.
  Types of parameters 'value' and 'value' are incompatible.
    Type 'CommonDropdownValue' is not assignable to type 'SetStateAction<string>'.
      Type 'string[]' is not assignable to type 'SetStateAction<string>'.
        Type 'string[]' is not assignable to type 'string'.

179                                 onChange={setNewLabelValue}
                                    ~~~~~~~~

  widgets/common/src/labels/CommonDropdown.tsx:11:5
    11     onChange: CommonDropdownOnChange;
           ~~~~~~~~
    The expected type comes from property 'onChange' which is declared here on type 'IntrinsicAttributes & ValueDropdownProps & { children?: ReactNode; }'


Found 4 errors.
```
</details>

File: widgets/common/src/labels/CommonDropdown.tsx [hunk 8]
```diff
@@ -1,8 +1,39 @@
...
+interface CommonDropdownProps extends KeyAndValueDropdownProps {
+    additionLabel: ComponentProps<typeof Dropdown>['additionLabel'];
+    disabled?: ComponentProps<typeof Dropdown>['disabled'];
+    fetchUrl: string;
+    name: ComponentProps<typeof Dropdown>['name'];
+    multiple?: ComponentProps<typeof Dropdown>['multiple'];
+    noResultsMessage?: ComponentProps<typeof Dropdown>['noResultsMessage'];
+    placeholder: ComponentProps<typeof Dropdown>['placeholder'];
+    tabIndex: ComponentProps<typeof Dropdown>['tabIndex'];
+    value: CommonDropdownValue;
```
[the remark this answers is not in the document]
Comment (qooban) [reply widgets/common/src/labels/CommonDropdown.tsx]: Yep, right, changed.

File: widgets/common/src/labels/LabelsInput.tsx [hunk 9]
```diff
@@ -35,11 +37,27 @@ function useReservedKeys(toolbox) {
...
+    toolbox
+}) => {
     const { useEffect, useRef } = React;
     const {
-        Basic: { Divider, Form, Icon, Segment },
-        Common: { DeploymentActions, RevertToDefaultIcon },
+        Basic: { Divider, Form, Icon, Segment, UnsafelyTypedFormField },
+        Common: {
+            DeploymentActions,
+            // @ts-ignore Property 'RevertToDefaultIcon' does not exist on type 'typeof Common'
```
[the remark this answers is not in the document]
Comment (qooban) [reply widgets/common/src/labels/LabelsInput.tsx]: Applied your suggestion to all places with a comment _"Property '*' does not exist on type 'typeof Common'"_

File: widgets/common/src/labels/LabelsList.tsx [hunk 10]
```diff
@@ -1,11 +1,18 @@
...
 import { sortLabels } from './common';
 
 const newLabelColor = 'blue';
 const labelLinesVisibleWithoutScroll = 6;
 const maxListHeight = `${labelLinesVisibleWithoutScroll * 2 + 0.2}em`;
 
-export default function LabelsList({ labels, onChange }) {
-    const { Label, Icon } = Stage.Basic;
+interface LabelsListProps {
+    labels: Label[];
```
[the remark this answers is not in the document]
Comment (qooban) [reply widgets/common/src/labels/LabelsList.tsx:10]: Good catch, but I actually did it the other way - I removed default value from that prop, which I think is more accurate.

File: widgets/common/src/labels/ManageLabelsModal.tsx (hunk 2 above)
Comment (qooban) [reply widgets/common/src/labels/ManageLabelsModal.tsx:4]: Thx. Slowly getting familiar with TS 🙂

File: widgets/common/src/hooks/useLabelInput.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Gelio) [reply widgets/common/src/hooks/useLabelInput.ts]: I appreciate the extra effort :+1:

File: widgets/common/src/labels/CommonDropdown.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (Gelio) [reply widgets/common/src/labels/CommonDropdown.tsx:6]: I see :smile: Thanks for trying it out and posting the log :+1: After thinking about it, those are indeed different. `(value: CommonDropdownValue) => void` forces consumers of the component to **have to handle** both `string` and `string[]` in the callbacks, whereas the current implementation does not.

We could use some TS magic to change the type of the `value` parameter based on whether `multiple` is set to `true` or not, but that's well out of scope here :smile: I'm just saying that would be possible

Let's leave it as is :slightly_smiling_face:

File: widgets/common/src/labels/LabelsList.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (Gelio) [reply widgets/common/src/labels/LabelsList.tsx:10]: Makes sense :slightly_smiling_face:

**Review by Gelio — APPROVED** [review]
Looks good to me :+1: 

Btw. I have noticed that the `6.0.0-build` branch is not **protected** in GitHub, meaning that we can merge to it without the CI passing / an approval. Probably we could also force push to it. Sounds unsafe to me :smile: I'll ask Yuval to make it protected on all repos :slightly_smiling_face:

**Pull request merged**